Parse arbitrary precision integers ..

hhugo opened the pull request:
.. and allow any char of [A-Za-z] as modifier for ints (previously 'l','L','n') and floats. 
This give more freedom to ppx rewritters (what about a ppx for zarith or literal for complex)

Checks are performed when translating from Parsetree to Typedtree.
Unknow_literal is raised if the modifier is not recognized ([lLn]?)
Integer_overflow is raised as before.

hhugo committed:
Parse arbitrary precision integers ..
.. and allow any identchar as modifier (previously 'l','L','n')
Also allow modifier for floats
This give more freedom to ppx rewritters (what about a ppx for zarith)

Checks are performed when translating from Parsetree to Typedtree.
Invalid_literal is raised if the modifier is not recognized ([lLn]?)
Integer_overflow is raised as before.

Lexer: use a-zA-Z for integer literal modifier

Lexer: Allow modifier on float

Clean wrt previous commits

Lexer: use named substring

Cleanup

typo

doc

fix after rebase

rebase on  trunk

Update typecore.ml

Fix printast.ml

hhugo commented on parsing/parser.mly:
```diff
@@ -2096,7 +2079,7 @@ class_longident:
     SHARP ident                 { Ptop_dir($2, Pdir_none) }
   | SHARP ident STRING          { Ptop_dir($2, Pdir_string (fst $3)) }
-  | SHARP ident INT             { Ptop_dir($2, Pdir_int $3) }
+  | SHARP ident INT             { Ptop_dir($2, Pdir_int (int_of_string $3)) }
```
This may fail

hhugo commented on parsing/lexer.mll:
```diff
@@ -330,31 +318,11 @@ rule token = parse
-          raise (Error(Literal_overflow "int", Location.curr lexbuf))
-      }
+  | int_literal { INT (Lexing.lexeme lexbuf) }
+  | (int_literal as num) (identchar + as modifier)
```
a 'modifier' could simply be ['a'-'z''A'-Z']

Octachron commented:
This patch looks really interesting. Would it  not make sense to have the same kind of support for float litteral? I can see some applications for qualified float litterals: complex number, single or quadruple precision litteral for instance. Most of these applications are not general enough to be supported within the compiler but would make nice ppx extensions.

hhugo commented on parsing/printast.ml:
```diff
@@ -887,7 +886,8 @@ and directive_argument i ppf x =
   match x with
   | Pdir_none -> line i ppf "Pdir_none\n"
   | Pdir_string (s) -> line i ppf "Pdir_string \"%s\"\n" s;
-  | Pdir_int (i) -> line i ppf "Pdir_int %d\n" i;
```
There was an error here because the argument `i` of the function was shadowed by the pattern matching,

hhugo replied:
15 years old typo !!

hhugo commented:
@Octachron, I've updated the PR to add modifier on floats.

samoht commented:
It could maybe be helpful if you add a few tests exercising the new code?

Drup commented:
I like the patch. I read the implementation and I think it's good.

I agree with @bobot on the minus sign.

I think there should be a function from Parsetree.constant to Asttypes.constant that do not need the typing env. It would be very useful for ppx writer. Returning an option would work well.

xavierleroy commented on parsing/lexer.mll:
```diff
@@ -330,29 +315,21 @@ rule token = parse
-          raise (Error(Literal_overflow "int", Location.curr lexbuf))
+  | int_literal { INT (Lexing.lexeme lexbuf, None) }
+  | int_literal literal_modifier
+      {
```
This would really benefit from using ocamllex's named sub-strings:

| (int_literal as lit) (literal_modifier as modif) { INT (lit, Some modif) }

hhugo commented:
I would prefer to keep the whole literal (minus sign included) inside the string. In most cases, it does not add any string processing as we already parse every int literal as negative numbers. (prepending  `"-"` if not  already negative, https://github.com/ocaml/ocaml/pull/170/files#diff-4408a3671b05f01d8b0dbe9df336e08bL173)

hhugo commented:
Do not merge until camlp4 PR is ready https://github.com/ocaml/camlp4/pull/90

hhugo commented:
PR ready for camlp4 https://github.com/ocaml/camlp4/pull/91

gasche commented:
This one looks ready to merge. What is the status of ocaml/camlp4#90 ?

hhugo commented:
Rebased on top of trunk.
Ready to merge unless one wants to add tests as proposed by @samoht

gasche commented:
Tests are always good. I'd love tests!

gasche commented on parsing/printast.ml:
```diff
@@ -51,15 +51,14 @@ let fmt_string_loc f x =
-  | Const_char (c) -> fprintf f "Const_char %02x" (Char.code c);
-  | Const_string (s, None) -> fprintf f "Const_string(%S,None)" s;
-  | Const_string (s, Some delim) ->
+  | PConst_int (i,None) -> fprintf f "Const_int %s" i;
```
Given that the format here is apparently meant to give re-parseability, you should reprint `PConst` rather than `Const`.

gasche commented:
If you write tests, it would be nice to include one checking that `-1073741824` is still correctly parsed as an integer literal (This corresponds to min_int for 32 machines.), to avoid [the issue described by this comment in MPR#4210](http://caml.inria.fr/mantis/view.php?id=4210#c3921).

gasche commented:
I have one worry with this patch: there is an ambiguity (resolved by the longest-lexeme rule) between literal modifiers and hexadecimal literals. For example, `32f` would be parsed as a literal modifier, but `0x32f` is a valid hexadecimal literal.

Wouldn't it be safer (for now) to either restrict literal modifiers to the non-hexadecimal cases (but this wouldn't be consistent with `nlL` which are valid for hexadecimal literals), or to letters strictly after `f` or `F`?

hhugo committed:
Tests for constants with modifiers

add some tests

more tests

hhugo commented:
How safer would that be ?
It seams to be similar to `1.2e` and `1.2e3`

dbuenzli commented:
> .. and allow any char of [A-Za-z] as modifier for ints

This may break existing programs, right ?

gasche commented:
Not directly, in the sense that programs that relied on `123foo` to parse were made illegal by #155 . That said, #155 has not yet been part of a release, so both changes together will break some programs, yes.

bobzhang commented:
So this patch is divided into two parts:
1. make parsetree using string literals (more loyal to user input)
2. extending lexer to support arbitrary modifiers
   
   With regard to 1, can we make changes to Asttypes.constant directly instead of creating a new type, Asttypes.constant is also used by Lambda, I am currently working on a project compiling lambda to readable javascript in the module level (still in the early days http://bobzhang.github.io/js-demo/), we could benefit such changes as well.

hhugo commented:
I've just rebased this PR onto trunk. Are people happy the current state ?

@bobzhang, you proposal of updating `Asttypes.constant` and not introducing a new type does not really make sense to me. One might still want to propagate constant as string to lambda but I would not do it inside this PR.

bobzhang commented:
my 2 cents: introducing another type actually increases the complexity of the patch (you have to update typeast and untypeast), I would say updating `Asttypes.constant` is good for both

hhugo commented:
The two types have very distinct (/ opposite )  purposes.
- `Parsetree.constant` have very little constraint to give more freedom to ppx rewriters
- `Asttypes.constant` enforce that constants are well-formed (integer should be int, int32, int64 .., and should not overflow).

Unifying the two types would make the patch more complex but feel free to prove me wrong

Drup commented:
As a ppx writer, I really like the distinction between the two types and I feel like it's useful.

damiendoligez commented:
@hhugo you need to add a line to `Changes`.

hhugo committed:
update Change log

GPR#170 can break existing programs
especially ppx rewriter as the Parsetree is updated

hhugo commented:
Change log updated.

note that https://github.com/hhugo/ocaml/commit/7681a73f5e66214fa0d701c91654bf5cfd8caa41 should _not_ be merged with the rest

damiendoligez commented on typing/typecore.ml:
```diff
@@ -4202,12 +4239,18 @@ let report_error env ppf = function
         "@[This form is not allowed as the type of the inlined record could \
          escape.@]"
-  | Unrefuted_pattern pat ->
+ | Unrefuted_pattern pat ->
```
Why did you change the indentation here? This looks wrong.

damiendoligez commented on Changes:
```diff
@@ -298,6 +298,9 @@ Features wishes:
   (Florian Angeletti)
 - GPR#252: improve build instructions in MSVC Windows README
   (Philip Daian)
+- GRP#170: Parse arbitrary precision integers and accept a single [A-Za-z]
```
typo on GPR

damiendoligez commented:
Good to merge once the typos are fixed.

hhugo commented:
@damiendoligez, should I rebase and remove the temporary commit related to `opam pin` ?

gasche commented:
Yes, rebasing is always a good idea to keep the commit history clean (we _do_ use bisect etc.), see [Contributing: Clean patch series](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#clean-patch-series).

hhugo commented:
PR rebased and cleared.
Note that the travis will fail because camlp4 is not pinned anymore.

gasche commented on Changes:
```diff
@@ -328,6 +328,9 @@ Features wishes:
   (Florian Angeletti)
 - GPR#252: improve build instructions in MSVC Windows README
   (Philip Daian)
+* GPR#170: Parse arbitrary precision integers and accept a single [A-Za-z]
```
Could you maybe add a line to clarify why this can break existing programs? It would probably suffice to just show an example: `let a = 1 in (+) 123a` was accepted but is now rejected. This would reassure users that, unless they did horrendous stuff, they should be quite safe.

(You should feel free to squash your Changelog commits together.)

gasche commented:
This looks good to merge, but I'm still mildly non-plussed by the interaction between modifiers in the a-f range and hexa literals. I assigned @damiendoligez so that he takes care of merging.

hhugo commented:
@gasche, thinking more about it, I now agree with your concern.
How do you feel about restricting the range for modifiers to ['g'-'z''G'-'Z'] ?

gasche commented:
I think that this would be a good, safe thing to do -- we can always extend the feature later with more experience.

damiendoligez commented:
Merged with [g-zG-Z].

alainfrisch commented:
The new test min_int.ml cannot be compiled on 32-bit architectures.  I've disabled it for now.

If someone insists on keeping it, please tweak the test suite to exclude this test (or create the 32-bit counterpart).

gasche commented:
We should have a warning on integer literals that do not fit on 32 machines, in the same spirit as the marshalling flag.

alainfrisch commented:
Note that there is already "ocamlc -compat-32", which triggers an error when the bytecode is not compatible with 32-bit.

damiendoligez commented:
I've "fixed" the test. Instead of adding heavy shell plumbing to compile a different source depending on the platform, I just changed it to the 32-bit `min_int`. This means the test is mostly useless on 64-bit, but the 32-bit platforms will test the corner case.

damiendoligez commented:
@hhugo [camlp4](https://github.com/ocaml/camlp4) is broken by this change. Can you have a look?

bobzhang commented:
curious do we have any ppx make use of such extension

Drup commented:
@bobzhang https://github.com/ocaml/Zarith/pull/4
